src/game/Maps/PathFinder.cpp

```diff
@@ -1048,12 +1048,13 @@ bool PathInfo::UpdateForMelee(Unit* pTarget, float meleeReach)
     // We have always keep at least 2 points (else, there is no mvt !)
     for (uint32 i = 1; i <= maxIndex; ++i)
     {
-        if (pTarget->IsWithinDist3d(m_pathPoints[i].x, m_pathPoints[i].y, m_pathPoints[i].z, meleeReach))
+        if (pTarget->IsWithinDist3d(m_pathPoints[i].x, m_pathPoints[i].y, m_pathPoints[i].z, meleeReach, SizeFactor::None))
         {
             Vector3 dirVect;
             pTarget->GetPosition(dirVect.x, dirVect.y, dirVect.z);
             dirVect -= m_pathPoints[i - 1];
-            float targetDist = pTarget->GetDistance(m_pathPoints[i - 1].x, m_pathPoints[i - 1].y, m_pathPoints[i - 1].z) - meleeReach;
+            float targetDist = pTarget->GetDistance(m_pathPoints[i - 1].x, m_pathPoints[i - 1].y, m_pathPoints[i - 1].z, SizeFactor::None);
+            targetDist -= meleeReach;
```

Comment:
Shouldn't you have so guards against `targetDist` becoming negative?

Reply:
It can be negative only if m_pathPoints[0] (the start pos of the movement) is inside the target's meleeReach. Normally the CanReachWithMeleeAutoAttack check that happens before should catch that and return to make the unit charge in place.

https://github.com/vmangos/core/assets/111737968/e3b2447a-e9ab-47db-9972-9334575fc1e4


https://github.com/vmangos/core/assets/111737968/a9b2dd6e-635b-42b8-9812-f94b5311682f

I tested removing the CanReachWithMeleeAutoAttack check and charging from inside melee range and this is what happens:

https://github.com/vmangos/core/assets/111737968/4baa1218-c114-4a75-b678-5a56a1ca9942

https://github.com/vmangos/core/assets/111737968/04a798b6-46f3-46ca-ab7f-75ad885b0c93

Nevertheless I suppose it's safer to check a second time and interrupt the charge to make double sure the charger doesn't path away from the target?

Though I'm curious to know what exactly do mobs in classic do when they charge a unit that is already in their melee range. All the examples I built this on is charging from a distance. In the case of players all charges have a 8 yard minimum range so it's impossible to make an analogy to mobs.